src/vs/platform/undoRedo/common/undoRedoService.ts

```diff
@@ -604,7 +604,8 @@ export class UndoRedoService implements IUndoRedoService {
 
 	public removeElements(resource: URI | string): void {
 		const strResource = typeof resource === 'string' ? resource : this.getUriComparisonKey(resource);
-		if (this._editStacks.has(strResource)) {
+		const sharesUndoRedoStack = typeof resource === 'string' ? false : this.getUriComparisonKey(resource) !== resource.toString();
+		if (this._editStacks.has(strResource) && !sharesUndoRedoStack) {
```

Comment:
I would suggest to move this check out of `IUndoRedoService`, and into the problematic caller of `IUndoRedoService.removeElements`, maybe at [baseCellViewModel.ts#L587](https://github.com/microsoft/vscode/blob/a92a03261cda3e0e32d8cb44e55dab4d40d165ea/src/vs/workbench/contrib/notebook/browser/viewModel/baseCellViewModel.ts#L587). Maybe that call should just be deleted, it doesn't seem reasonable to clear the undo redo elements when a single cell gets disposed. That call was added relatively recently via https://github.com/microsoft/vscode/commit/1d9ac5e14a9ee2bcc1dc26cb69f87eb5af0bd4b3 .

Adding this condition inside the `IUndoRedoService` would make it possible that undo redo elements would leak forever in certain conditions, depending on the `uriComparisonKey` implementation.

Similarly, the check in `ModelService._onWillDispose` takes care to avoid calling `IUndoRedoService.removeElements` when a TextModel instantiated for a cell is disposed, because it would remove undo elements for the entire notebook.

Reply:
@alexdima Sounds good thanks much for the review, I'd noted in my comment that the baseCellViewModel.ts was the other spot I was considering making this check. So moving it there sounds great, as noted it would restrict it to just the notebook scenario when we know that the full document will eventually get disposed. So we don't have to worry about a leak there. My first change is more general and I could see it create a leak in a non-notebook scenario. Good call to make it safer. 

From my understanding the call to removeElement there is still needed in the scenario where perCell undo / redo setting is turned on. In that case the cell uri maps to itself instead of the document, and each cell has its own undo / redo stack. So I think that I'll just gate the call to removeElement in the cell view model on the same uriComparisonKey check.